### PR TITLE
Adding maximum restart attempts to BackoffSupervisor #24769

### PR DESCRIPTION
**Pull request by salehkhazaei**
Refs #24769

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @salehkhazaei,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by lightbend-cla-validator** [comment]
Hi @salehkhazaei,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by salehkhazaei** [comment]
I've signed the CLA.

**Comment by ktoso** [comment]
OK TO TEST

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 1]
```diff
@@ -71,14 +71,17 @@ object Backoff {
...
-    minBackoff:   FiniteDuration,
-    maxBackoff:   FiniteDuration,
-    randomFactor: Double): BackoffOptions =
-    BackoffOptionsImpl(RestartImpliesFailure, childProps, childName, minBackoff, maxBackoff, randomFactor)
+    childProps:         Props,
+    childName:          String,
+    minBackoff:         FiniteDuration,
+    maxBackoff:         FiniteDuration,
+    randomFactor:       Double,
+    maxRestartAttempts: Int): BackoffOptions =
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Unfortunately 'just' adding an additional parameter here will break backwards compatibility, so you'd need to add an extra (deprecated) `onFailure` method with the old signature.

**Review by raboof — COMMENTED** [review]
The build is failing due to backwards compatibility

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 2]
```diff
@@ -318,6 +330,7 @@ private final case class BackoffOptionsImpl(
   minBackoff:         FiniteDuration,
```
Comment (salehkhazaei) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Mima reports binary compatibility issue with BackoffOptionsImpl class.
Since the class is private and used only in BackoffOptions, I'm concluding that changing it shouldn't raise a binary compatibility issue and it can be ignored. am I right?

Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: If I'm wrong, I would appreciate if you could help me on how to fix this one.

Comment (raboof) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Indeed changes to `BackoffOptionsImpl` should be fine, so you can exclude those from MiMa as described in the CONTRIBUTING.md. It looks like you have some similar problems in `Backoff`, though, which do need to be addressed. Does that help?

Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Yes, it does. Thanks alot.

File: akka-actor-tests/src/test/scala/akka/pattern/BackoffOnRestartSupervisorSpec.scala [hunk 3]
```diff
@@ -52,7 +52,7 @@ class TestParentActor(probe: ActorRef, supervisorProps: Props) extends Actor {
 class BackoffOnRestartSupervisorSpec extends AkkaSpec with ImplicitSender {
 
   def supervisorProps(probeRef: ActorRef) = {
-    val options = Backoff.onFailure(TestActor.props(probeRef), "someChildName", 200 millis, 10 seconds, 0.0)
+    val options = Backoff.onFailure(TestActor.props(probeRef), "someChildName", 200 millis, 10 seconds, 0.0, -1)
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/pattern/BackoffOnRestartSupervisorSpec.scala]: Even if private/test, perhaps consider using a named parameter here (`maxRestartAttempts = -1`) to make it easier to read.

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 4]
```diff
@@ -125,14 +183,136 @@ object Backoff {
...
+   *   will be started and supervised
+   * @param childName name of the child actor
+   * @param minBackoff minimum (initial) duration until the child actor will
+   *   started again, if it is terminated
+   * @param maxBackoff the exponential back-off is capped to this duration
+   * @param randomFactor after calculation of the exponential back-off an additional
+   *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
+   *   In order to skip this additional delay pass in `0`.
+   */
+  @deprecated
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Since this is the Java API, use the Java `@Deprecated`

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 5]
```diff
@@ -125,14 +183,136 @@ object Backoff {
    * @param randomFactor after calculation of the exponential back-off an additional
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
+   * @param maxRestartAttempts maximum number of attempts to restart the child actor.
+   *   In order to restart infinitely pass in `-1`
    */
+  def onFailure(
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Might be nice to mark this 'Java API' in the scaladoc while we're here

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 6]
```diff
@@ -248,13 +493,14 @@ object Backoff {
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
    */
+  @deprecated
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: (Java-style deprecation here as well)

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 7]
```diff
@@ -71,7 +71,65 @@ object Backoff {
...
+   *   will be started and supervised
+   * @param childName name of the child actor
+   * @param minBackoff minimum (initial) duration until the child actor will
+   *   started again, if it is terminated
+   * @param maxBackoff the exponential back-off is capped to this duration
+   * @param randomFactor after calculation of the exponential back-off an additional
+   *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
+   *   In order to skip this additional delay pass in `0`.
+   */
+  @deprecated
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: add 'message' and 'since' parameters

**Review by raboof — COMMENTED** [review]
Starting to shape up nicely! Added some comments here and there.

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala (hunk 7 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: which version should I put in 'since'?

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 8]
```diff
@@ -71,7 +71,65 @@ object Backoff {
    * @param randomFactor after calculation of the exponential back-off an additional
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
+   * @param maxRestartAttempts maximum number of attempts to restart the child actor.
+   *   In order to restart infinitely pass in `-1`
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Describe what happens when the max has been reached.

File: akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala [hunk 9]
```diff
@@ -67,9 +80,11 @@ private class BackoffOnRestartSupervisor(
     case Terminated(`childRef`) ⇒
       become(receive)
       child = None
-      val restartDelay = BackoffSupervisor.calculateDelay(restartCount, minBackoff, maxBackoff, randomFactor)
-      context.system.scheduler.scheduleOnce(restartDelay, self, BackoffSupervisor.StartChild)
-      restartCount += 1
+      if (maxRestartAttempts == -1 || restartCount < maxRestartAttempts) {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: how is this related to `if (strategy.maxNrOfRetries >= 0 && nextRestartCount > strategy.maxNrOfRetries)` a few lines up?

Do we have two max values in play now?

File: akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala [hunk 10]
```diff
@@ -67,9 +80,11 @@ private class BackoffOnRestartSupervisor(
...
+        val restartDelay = BackoffSupervisor.calculateDelay(restartCount, minBackoff, maxBackoff, randomFactor)
+        context.system.scheduler.scheduleOnce(restartDelay, self, BackoffSupervisor.StartChild)
+        restartCount += 1
+      }
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: what should happen when the max has been reached? Logging and `stop(self)`?

File: akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala (hunk 9 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: I've realized it's existence before, but changing the maxNrOfRetries of the strategy which has been set was a little confusing to me. I should've asked first. What do you propose? should I rewrite it using the maxNrOfRetries of the strategy?

File: akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala (hunk 10 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: I'll add these two.

File: akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala (hunk 9 above)
Comment (patriknw) [reply akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: I think it can be confusing to have two different competing max. Could we take a step back and list all API methods and see where we should add it.

* BackoffSupervisor.props - here we could add an overloaded method that takes the additional maxNrOfRetries
* BackoffSupervisor.propsWithSupervisorStrategy - if the strategy is given then we should not have another parameter for maxNrOfRetries, instead that is taken from the strategy
* Backoff.onFailure - add an overloaded method that takes the additional maxNrOfRetries
* Backoff.onStop - add an overloaded method that takes the additional maxNrOfRetries
* BackoffOptions.withSupervisorStrategy - if this is used then it will replace previously defined maxNrOfRetries if any (must be documented)
* BackoffOptions.withDefaultStoppingStrategy - don't replace previously defined maxNrOfRetries if any

Did I miss any other way to create the backoff supervisor?

Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: I think that's all. 
Thank you.

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala (hunk 5 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Done

**Comment by salehkhazaei** [comment]
Failure of tests was due to timeout on irrelevant tests. Should I be worried or it is just due to jenkins being too busy?

File: akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala (hunk 9 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala]: I have implemented it the way we discussed. Could you review it again?

**Comment by raboof** [comment]
PLS BUILD

**Comment by raboof** [comment]
Hm, it failed again and I can reproduce it locally as well. Could you look into it a bit closer?

**Comment by salehkhazaei** [comment]
Thank you, I'll look into it.

File: akka-actor/src/main/mima-filters/2.5.14.backwards.excludes [hunk 11]
```diff
@@ -1,3 +1,6 @@
+#25435 - Adding maximum restart attempts to BackoffSupervisor
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.pattern.BackoffOptions.withMaxNrOfRetries")
```
Comment (salehkhazaei) [inline akka-actor/src/main/mima-filters/2.5.14.backwards.excludes]: I wasn't sure if it's safe to exclude this binary compatibility issue. Could you take a look?
@raboof @ktoso

**Review by patriknw — COMMENTED** [review]
Looking good, great work. Some minor...

File: akka-actor/src/main/scala/akka/actor/FaultHandling.scala [hunk 12]
```diff
@@ -170,6 +170,15 @@ object SupervisorStrategy extends SupervisorStrategyLowPriorityImplicits {
     OneForOneStrategy()(defaultDecider)
   }
 
+  /**
+   * default strategy with maximum number of retries.
+   * @param maxNrOfRetries the number of times a child actor is allowed to be restarted, negative value means no limit,
+   *  if the limit is exceeded the child actor is stopped
+   */
+  def defaultStrategyWithMaxNrOfRetries(maxNrOfRetries: Int): SupervisorStrategy = {
```
[the remark this answers is not in the document]
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/actor/FaultHandling.scala]: I can remove this function and use below code in BackOffSupervisor. Is it good?

`defaultStrategy match {
case oneForOne: OneForOneStrategy => oneForOne.withMaxNrOfRetries(maxNrOfRetries)
case s => s
}`

File: akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala [hunk 13]
```diff
@@ -242,5 +242,96 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender {
...
+        supervisor ! BackoffSupervisor.GetCurrentChild
+        val c1 = expectMsgType[BackoffSupervisor.CurrentChild].ref.get
+        watch(c1)
+        c1 ! "boom"
+        expectTerminated(c1)
+
+        supervisor ! BackoffSupervisor.GetRestartCount
+        expectMsg(BackoffSupervisor.RestartCount(1))
+
+        val c2 = waitForChild().get
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala]: Perhaps you could use an `eventually` block instead of the custom `waitForChild` here?

File: akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala [hunk 14]
```diff
@@ -275,9 +333,21 @@ final class BackoffSupervisor(
   def onTerminated: Receive = {
     case Terminated(ref) if child.contains(ref) ⇒
       child = None
-      val restartDelay = calculateDelay(restartCount, minBackoff, maxBackoff, randomFactor)
-      context.system.scheduler.scheduleOnce(restartDelay, self, StartChild)
-      restartCount += 1
+      val maxNrOfRetries = strategy match {
+        case oneForOne: OneForOneStrategy ⇒ oneForOne.maxNrOfRetries
+        case _                            ⇒ -1
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala]: I wonder if we should create a constant instead of using the magic `-1` number - might be OK like this though.

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala [hunk 15]
```diff
@@ -187,6 +371,7 @@ object Backoff {
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
    */
+  @deprecated("Use the overloaded one which accepts maxNrOfRetries instead.", "2.5.14")
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: 2.5.14 should be updated here

**Review by raboof — APPROVED** [review]
Thanks for keeping at this, looks good now I think!

File: akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala (hunk 15 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala]: Done

File: akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala (hunk 13 above)
Comment (salehkhazaei) [reply akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala]: Done

File: akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala (hunk 14 above)
Comment (salehkhazaei) [reply akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala]: I thought of that too. but since -1 was used in other places and it's meaning was mentioned in the comments, I didn't use a constant. So, should I change it?

**Review by patriknw — APPROVED** [review]
LGTM, thanks @salehkhazaei
